mkosi/__init__.py

```diff
@@ -5588,24 +5664,86 @@ def make_build_dir(args: CommandLineArguments) -> None:
     mkdir_last(args.build_dir, 0o755)
 
 
+def setup_ssh(args: CommandLineArguments, root: str, do_run_build_script: bool, for_cache: bool) -> Optional[TextIO]:
+    if do_run_build_script or for_cache or not args.ssh:
+        return None
+
+    if args.distribution in [Distribution.ubuntu, Distribution.debian]:
+        unit = "ssh"
+    else:
+        unit = "sshd"
+
+    run(["systemctl", "--root", root, "enable", unit])
+
+    f: TextIO = cast(
+        TextIO,
+        tempfile.NamedTemporaryFile(
+            mode="w+", prefix=".mkosi-", encoding="utf-8", dir=os.path.dirname(args.output_sshkey)
+        ),
+    )
+
+    with complete_step("Generating SSH keypair"):
+        # Write a 'y' to confirm to overwrite the file.
+        run(["ssh-keygen", "-f", f.name, "-N", args.password or ""], input=f"y\n", text=True)
+
+    copy_file(f"{f.name}.pub", os.path.join(root, "root/.ssh/authorized_keys"))
+    os.remove(f"{f.name}.pub")
+
+    os.chmod(os.path.join(root, "root/.ssh/authorized_keys"), 0o600)
+
+    return f
+
+
+def setup_network_veth(args: CommandLineArguments, root: str, do_run_build_script: bool, for_cache: bool) -> None:
+    if do_run_build_script or for_cache or not args.network_veth:
+        return
+
+    network_file = os.path.join(root, "etc/systemd/network/80-mkosi-network-veth.network")
+    with open(network_file, "w") as f:
+        f.write(
+            dedent(
+                """\
+                [Match]
+                Virtualization=qemu
+                Type=ether
+
+                [Network]
+                DHCP=yes
+                LinkLocalAddressing=yes
+                LLDP=yes
+                EmitLLDP=customer-bridge
+
+                [DHCP]
+                UseTimezone=yes
+                """
+            )
+        )
+
+    os.chmod(network_file, 0o644)
+
+    run(["systemctl", "--root", root, "enable", "systemd-networkd"])
```

Comment:
at least on Debian systems, systemd-resolved also need to be enabled as it's not by default

Reply:
We don't need resolved anymore to get --ssh working though with the current approach.

We could enable resolved by default when using --network-veth but let's put that in a separate PR.

Reply:
Yeah I didn't mean it for ssh, but in general for networking. But yes it can be done separately